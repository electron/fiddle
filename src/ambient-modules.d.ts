declare module 'namor';
declare module 'shell-path';
