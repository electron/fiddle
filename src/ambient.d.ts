import * as MonacoType from 'monaco-editor';

import { App } from './renderer/app';

declare global {
  interface Window {
    ElectronFiddle: {
      app: App;
      appPaths: Record<string, string>;
      monaco: typeof MonacoType;
    };
  }
}
