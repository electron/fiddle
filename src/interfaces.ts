import { InstallState } from '@vertedinde/fiddle-core';

export type Files = Map<string, string>;

export type FileTransform = (files: Files) => Promise<Files>;

export enum VersionSource {
  remote = 'remote',
  local = 'local',
}

export enum GistActionType {
  publish = 'Publish',
  update = 'Update',
  delete = 'Delete',
}

export enum GistActionState {
  publishing = 'publishing',
  updating = 'updating',
  deleting = 'deleting',
  none = 'none',
}

export interface Version {
  version: string;
  name?: string;
  localPath?: string;
  node?: string;
}

export enum RunResult {
  SUCCESS = 'success', // exit code === 0
  FAILURE = 'failure', // ran, but exit code !== 0
  INVALID = 'invalid', // could not run
}

export interface RunnableVersion extends Version {
  state: InstallState;
  source: VersionSource;
  downloadProgress?: number;
}

export const enum ElectronReleaseChannel {
  stable = 'Stable',
  beta = 'Beta',
  nightly = 'Nightly',
}

export interface SetFiddleOptions {
  filePath?: string;
  templateName?: string;
  gistId?: string;
}

export interface SetUpMenuOptions {
  acceleratorsToBlock?: BlockableAccelerator[] | null;
  activeTemplate?: string | null;
}

export interface SetupRequest {
  fiddle?: SetFiddleOptions;
  version?: string;
  showChannels: ElectronReleaseChannel[];
  hideChannels: ElectronReleaseChannel[];
  useObsolete?: boolean;
}

export interface BisectRequest {
  setup: SetupRequest;
  goodVersion: string;
  badVersion: string;
}

export interface TestRequest {
  setup: SetupRequest;
}

export interface OutputEntry {
  text: string;
  timeString: string;
  isNotPre?: boolean;
}

export interface OutputOptions {
  bypassBuffer?: boolean;
  isNotPre?: boolean;
}

export interface GenericDialogOptions {
  type: GenericDialogType;
  ok: string;
  cancel?: string;
  wantsInput: boolean;
  defaultInput?: string;
  label: string | JSX.Element;
  placeholder?: string;
}

export interface Contributor {
  url: string;
  api: string;
  login: string;
  avatar: string;
  name: string;
  bio: string;
  location: string;
}

export interface Templates {
  [index: string]: string | Templates;
}

export const enum GenericDialogType {
  'confirm' = 'confirm',
  'warning' = 'warning',
  'success' = 'success',
}

export type EditorId = `${string}.${'js' | 'html' | 'css'}`;

export type EditorValues = Record<EditorId, string>;

// main.js gets special treatment: it is required as the entry point
// when we run fiddles or create a package.json to package fiddles.
export const MAIN_JS = 'main.js';

export const PACKAGE_NAME = 'package.json';

export type ArrowPosition = 'top' | 'left' | 'bottom' | 'right';

export const enum BlockableAccelerator {
  save = 'save',
  saveAs = 'saveAs',
}
