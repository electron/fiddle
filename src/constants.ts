export const SENTRY_DSN =
  'https://966a5b01ac8d4941b81e4ebd0ab4c991@sentry.io/1882540';
