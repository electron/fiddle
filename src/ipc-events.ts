export enum IpcEvents {
  OPEN_SETTINGS = 'OPEN_SETTINGS',
  LOAD_GIST_REQUEST = 'LOAD_GIST_REQUEST',
  LOAD_ELECTRON_EXAMPLE_REQUEST = 'LOAD_ELECTRON_EXAMPLE_REQUEST',
  FIDDLE_RUN = 'FIDDLE_RUN',
  FIDDLE_PACKAGE = 'FIDDLE_PACKAGE',
  FIDDLE_MAKE = 'FIDDLE_MAKE',
  MONACO_EXECUTE_COMMAND = 'MONACO_EXECUTE_COMMAND',
  MONACO_TOGGLE_OPTION = 'MONACO_TOGGLE_OPTION',
  FS_NEW_FIDDLE = 'FS_NEW_FIDDLE',
  FS_NEW_TEST = 'FS_NEW_TEST',
  FS_OPEN_FIDDLE = 'FS_OPEN_FIDDLE',
  FS_OPEN_TEMPLATE = 'FS_OPEN_TEMPLATE',
  FS_SAVE_FIDDLE = 'FS_SAVE_FIDDLE',
  FS_SAVE_FIDDLE_AS = 'FS_SAVE_FIDDLE_AS',
  FS_SAVE_FIDDLE_GIST = 'FS_SAVE_FIDDLE_GIST',
  FS_SAVE_FIDDLE_FORGE = 'FS_SAVE_FIDDLE_FORGE',
  FS_SAVE_FIDDLE_DIALOG = 'FS_SAVE_FIDDLE_DIALOG',
  FS_SAVE_FIDDLE_ERROR = 'FS_SAVE_FIDDLE_ERROR',
  SHOW_WARNING_DIALOG = 'SHOW_WARNING_DIALOG',
  SHOW_CONFIRMATION_DIALOG = 'SHOW_CONFIRMATION_DIALOG',
  SHOW_WELCOME_TOUR = 'SHOW_WELCOME_TOUR',
  CLEAR_CONSOLE = 'CLEAR_CONSOLE',
  LOAD_LOCAL_VERSION_FOLDER = 'LOAD_LOCAL_VERSION_FOLDER',
  SHOW_LOCAL_VERSION_FOLDER_DIALOG = 'SHOW_LOCAL_VERSION_FOLDER_DIALOG',
  BISECT_COMMANDS_TOGGLE = 'BISECT_COMMANDS_TOGGLE',
  SHOW_INACTIVE = 'SHOW_INACTIVE',
  BEFORE_QUIT = 'BEFORE_QUIT',
  CONFIRM_QUIT = 'CONFIRM_QUIT',
  GET_APP_PATHS = 'GET_APP_PATHS',
  SELECT_ALL_IN_EDITOR = 'SELECT_ALL_IN_EDITOR',
  BLOCK_ACCELERATORS = 'BLOCK_ACCELERATORS',
  SET_SHOW_ME_TEMPLATE = 'SET_SHOW_ME_TEMPLATE',
  CLICK_TITLEBAR_MAC = 'CLICK_TITLEBAR_MAC',
  TASK_BISECT = 'TASK_BISECT',
  TASK_TEST = 'TASK_TEST',
  TASK_DONE = 'TASK_DONE',
  OUTPUT_ENTRY = 'OUTPUT_ENTRY',
  RELOAD_WINDOW = 'RELOAD_WINDOW',
}

export const ipcMainEvents = [
  IpcEvents.FS_SAVE_FIDDLE_DIALOG,
  IpcEvents.FS_SAVE_FIDDLE,
  IpcEvents.SHOW_WARNING_DIALOG,
  IpcEvents.SHOW_CONFIRMATION_DIALOG,
  IpcEvents.SHOW_LOCAL_VERSION_FOLDER_DIALOG,
  IpcEvents.SHOW_INACTIVE,
  IpcEvents.CONFIRM_QUIT,
  IpcEvents.SET_SHOW_ME_TEMPLATE,
  IpcEvents.BLOCK_ACCELERATORS,
  IpcEvents.CLICK_TITLEBAR_MAC,
  IpcEvents.OUTPUT_ENTRY,
  IpcEvents.TASK_DONE,
  IpcEvents.RELOAD_WINDOW,
];

export const ipcRendererEvents = [
  IpcEvents.OPEN_SETTINGS,
  IpcEvents.LOAD_GIST_REQUEST,
  IpcEvents.LOAD_ELECTRON_EXAMPLE_REQUEST,
  IpcEvents.FIDDLE_RUN,
  IpcEvents.FIDDLE_PACKAGE,
  IpcEvents.FIDDLE_MAKE,
  IpcEvents.MONACO_EXECUTE_COMMAND,
  IpcEvents.MONACO_TOGGLE_OPTION,
  IpcEvents.FS_NEW_FIDDLE,
  IpcEvents.FS_NEW_TEST,
  IpcEvents.FS_OPEN_FIDDLE,
  IpcEvents.FS_OPEN_TEMPLATE,
  IpcEvents.FS_SAVE_FIDDLE,
  IpcEvents.FS_SAVE_FIDDLE_AS,
  IpcEvents.FS_SAVE_FIDDLE_FORGE,
  IpcEvents.FS_SAVE_FIDDLE_GIST,
  IpcEvents.SHOW_WELCOME_TOUR,
  IpcEvents.CLEAR_CONSOLE,
  IpcEvents.LOAD_LOCAL_VERSION_FOLDER,
  IpcEvents.BISECT_COMMANDS_TOGGLE,
  IpcEvents.BEFORE_QUIT,
  IpcEvents.GET_APP_PATHS,
  IpcEvents.SELECT_ALL_IN_EDITOR,
  IpcEvents.SET_SHOW_ME_TEMPLATE,
  IpcEvents.TASK_BISECT,
  IpcEvents.TASK_TEST,
];

export const WEBCONTENTS_READY_FOR_IPC_SIGNAL =
  'WEBCONTENTS_READY_FOR_IPC_SIGNAL';
