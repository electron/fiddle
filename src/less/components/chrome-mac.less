.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 24px;
    font-size: 1.4rem;
    line-height: 15px;

    small,
    p,
    span {
      user-select: none;
    }
  }

  .big_sur_or_later {
    height: 30px !important;
  }
}
