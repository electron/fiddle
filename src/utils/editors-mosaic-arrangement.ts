import { MosaicDirection, MosaicNode } from 'react-mosaic-component';

import { MosaicId } from '../interfaces';

/**
 * Create a mosaic arrangement given an array of editor ids.
 *
 * @export
 * @param {Array<MosaicId>} input
 * @returns {MosaicNode<MosaicId>}
 */
export function createMosaicArrangement(
  input: Array<MosaicId>,
  direction: MosaicDirection = 'row',
): MosaicNode<MosaicId> {
  if (input.length === 1) {
    return input[0];
  }

  // This cuts out the first half of input. Input becomes the second half.
  const secondHalf = [...input];
  const firstHalf = secondHalf.splice(0, Math.floor(secondHalf.length / 2));

  return {
    direction,
    first: createMosaicArrangement(firstHalf, 'column'),
    second: createMosaicArrangement(secondHalf, 'column'),
  };
}

/**
 * Returns an array of visible editors given a mosaic arrangement.
 *
 * @export
 * @param {MosaicNode<MosaicId> | null} input
 * @returns {Array<MosaicId>}
 */
export function getVisibleMosaics(
  input: MosaicNode<MosaicId> | null,
): Array<MosaicId> {
  // Handle the unlikely null case
  if (!input) return [];

  // Handle the case where only one editor is visible
  if (typeof input === 'string') {
    return [input];
  }

  // Handle the other cases (2 - 4)
  const result = [];
  for (const node of [input.first, input.second]) {
    if (typeof node === 'string') {
      result.push(node);
    } else {
      result.push(...getVisibleMosaics(node));
    }
  }

  return result;
}
