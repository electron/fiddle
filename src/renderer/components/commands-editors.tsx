import {
  Button,
  Menu,
  MenuDivider,
  MenuItem,
  Popover,
  Position,
} from '@blueprintjs/core';
import { when } from 'mobx';
import { observer } from 'mobx-react';
import * as React from 'react';

import {
  DEFAULT_EDITORS,
  DefaultEditorId,
  EditorId,
  GenericDialogType,
} from '../../interfaces';
import { AppState } from '../state';
import { getEditorTitle, isSupportedFile } from '../../utils/editor-utils';

interface EditorDropdownState {
  value: string;
}

interface EditorDropdownProps {
  appState: AppState;
}

/**
 * A dropdown allowing users to toggle the various editors
 *
 * @class EditorDropdown
 * @extends {React.Component<EditorDropdownProps, EditorDropdownState>}
 */
@observer
export class EditorDropdown extends React.Component<
  EditorDropdownProps,
  EditorDropdownState
> {
  constructor(props: EditorDropdownProps) {
    super(props);

    this.onItemClick = this.onItemClick.bind(this);
    this.showCustomEditorDialog = this.showCustomEditorDialog.bind(this);
    this.addCustomEditor = this.addCustomEditor.bind(this);
    this.removeCustomEditor = this.removeCustomEditor.bind(this);
  }

  public render() {
    return (
      <>
        <Popover content={this.renderMenu()} position={Position.BOTTOM}>
          <Button icon="applications" text="Editors" />
        </Popover>
      </>
    );
  }

  public renderMenu() {
    return <Menu>{...this.renderMenuItems()}</Menu>;
  }

  public renderMenuItems() {
    const { appState } = this.props;
    const result: Array<JSX.Element> = [];
    const visibleMosaics = appState.editorMosaic.getVisibleMosaics();

    const allEditors = [
      ...DEFAULT_EDITORS,
      ...appState.editorMosaic.customMosaics,
    ];
    for (const id of allEditors) {
      const icon = visibleMosaics.includes(id) ? 'eye-open' : 'eye-off';
      const title = getEditorTitle(id);

      if (!DEFAULT_EDITORS.includes(id as any)) {
        result.push(
          <MenuItem
            icon={icon}
            key={id}
            text={title}
            id={id}
            onClick={this.onItemClick}
            // Can't hide last editor panel.
            disabled={appState.editorMosaic.mosaicArrangement === id}
          >
            <MenuItem
              icon={'cross'}
              id={id}
              onClick={this.removeCustomEditor}
              text={'Remove'}
            />
          </MenuItem>,
        );
      } else {
        result.push(
          <MenuItem
            icon={icon}
            key={id}
            text={title}
            id={id}
            onClick={this.onItemClick}
            // Can't hide last editor panel.
            disabled={appState.editorMosaic.mosaicArrangement === id}
          />,
        );
      }
    }

    result.push(
      <React.Fragment key={'fragment-custom-editor'}>
        <MenuDivider />
        <MenuItem
          icon="plus"
          key="add-custom-editor"
          text="Add Custom Editor"
          onClick={this.addCustomEditor}
        />
      </React.Fragment>,
    );

    result.push(
      <React.Fragment key={'fragment-reset-layout'}>
        <MenuDivider />
        <MenuItem
          icon="grid-view"
          key="reset-layout"
          text="Reset Layout"
          onClick={appState.editorMosaic.resetEditorLayout}
        />
      </React.Fragment>,
    );

    return result;
  }

  public async showCustomEditorDialog() {
    const { appState } = this.props;

    appState.setGenericDialogOptions({
      type: GenericDialogType.confirm,
      label: 'Enter a filename for your custom editor',
      wantsInput: true,
      ok: 'Create',
      cancel: 'Cancel',
      placeholder: 'file.js',
    });

    appState.toggleGenericDialog();
    await when(() => !appState.isGenericDialogShowing);

    return {
      cancelled: !appState.genericDialogLastResult,
      result: appState.genericDialogLastInput,
    };
  }

  public async addCustomEditor() {
    const { appState } = this.props;

    const { cancelled, result } = await this.showCustomEditorDialog();

    if (cancelled) return;

    // Fail if editor name is not an accepted file type.
    if (!result || !isSupportedFile(result)) {
      appState.setGenericDialogOptions({
        type: GenericDialogType.warning,
        label:
          'Invalid custom editor name - must be either an html, js, or css file.',
        cancel: undefined,
      });

      appState.toggleGenericDialog();
    } else {
      const name = result as EditorId;

      // Also fail if the user tries to create two identical editors.
      if (
        appState.editorMosaic.customMosaics.includes(name) ||
        Object.values(DefaultEditorId).includes(name as DefaultEditorId)
      ) {
        appState.setGenericDialogOptions({
          type: GenericDialogType.warning,
          label: `Custom editor name ${name} already exists - duplicates are not allowed`,
          cancel: undefined,
        });

        appState.toggleGenericDialog();
      } else {
        appState.editorMosaic.customMosaics.push(name);
        appState.editorMosaic.showMosaic(name);
      }
    }
  }

  public async removeCustomEditor(event: React.MouseEvent) {
    const { id } = event.currentTarget;
    const { appState } = this.props;

    console.log(`EditorDropdown: Removing custom editor ${id}`);
    appState.editorMosaic.removeCustomMosaic(id as EditorId);
  }

  public onItemClick(event: React.MouseEvent) {
    const { id } = event.currentTarget;
    const { appState } = this.props;
    const { editorMosaic } = appState;
    const visibleMosaics = appState.editorMosaic.getVisibleMosaics();

    if (visibleMosaics.includes(id as EditorId)) {
      console.log(`EditorDropdown: Closing ${id}`);
      editorMosaic.hideAndBackupMosaic(id as EditorId);
    } else {
      console.log(`EditorDropdown: Opening ${id}`);
      editorMosaic.showMosaic(id as EditorId);
    }
  }
}
