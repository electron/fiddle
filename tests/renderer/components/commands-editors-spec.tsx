import { mount, shallow } from 'enzyme';
import * as React from 'react';

import { EditorId } from '../../../src/interfaces';
import { EditorDropdown } from '../../../src/renderer/components/commands-editors';
import { getVisibleMosaics } from '../../../src/utils/editors-mosaic-arrangement';

jest.mock('../../../src/utils/editors-mosaic-arrangement');

describe('EditorDropdown component', () => {
  let store: any;

  beforeEach(() => {
    (process.env as any).FIDDLE_DOCS_DEMOS = false;

    store = {
      hideAndBackupMosaic: jest.fn(),
      showMosaic: jest.fn(),
      closedPanels: {},
    };

    (getVisibleMosaics as jest.Mock).mockReturnValue([
      EditorId.html,
      EditorId.renderer,
    ]);
  });

  it('renders', () => {
    const wrapper = shallow(<EditorDropdown appState={store} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the extra button if the FIDDLE_DOCS_DEMOS is set', () => {
    (process.env as any).FIDDLE_DOCS_DEMOS = true;

    const wrapper = shallow(<EditorDropdown appState={store} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles a click for an item', () => {
    const wrapper = mount(<EditorDropdown appState={store} />);
    const dropdown = wrapper.instance() as EditorDropdown;

    dropdown.onItemClick({ currentTarget: { id: EditorId.html } } as any);
    expect(store.hideAndBackupMosaic).toHaveBeenCalledTimes(1);
    expect(store.showMosaic).toHaveBeenCalledTimes(0);

    dropdown.onItemClick({ currentTarget: { id: EditorId.main } } as any);
    expect(store.hideAndBackupMosaic).toHaveBeenCalledTimes(1);
    expect(store.showMosaic).toHaveBeenCalledTimes(1);
  });

  it('disables hide button if only one editor open', () => {
    store.mosaicArrangement = 'html';
    (getVisibleMosaics as jest.Mock).mockReturnValue([EditorId.html]);

    const wrapper = mount(<EditorDropdown appState={store} />);
    const instance = wrapper.instance() as EditorDropdown;
    const menu = instance.renderMenuItems();

    expect(menu).toMatchSnapshot();
  });
});
