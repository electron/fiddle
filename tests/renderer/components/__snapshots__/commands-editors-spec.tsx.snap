// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorDropdown component renders 1`] = `
<Fragment>
  <Blueprint3.Popover
    boundary="scrollParent"
    captureDismiss={false}
    content={
      <Blueprint3.Menu>
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="index.html"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="HTML (index.html)"
        />
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="renderer.js"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="Renderer Process (renderer.js)"
        />
        <React.Fragment>
          <Blueprint3.MenuDivider />
          <Blueprint3.MenuItem
            disabled={false}
            icon="plus"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Add Custom Editor"
          />
        </React.Fragment>
        <React.Fragment>
          <Blueprint3.MenuDivider />
          <Blueprint3.MenuItem
            disabled={false}
            icon="grid-view"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Reset Layout"
          />
        </React.Fragment>
      </Blueprint3.Menu>
    }
    defaultIsOpen={false}
    disabled={false}
    fill={false}
    hasBackdrop={false}
    hoverCloseDelay={300}
    hoverOpenDelay={150}
    inheritDarkTheme={true}
    interactionKind="click"
    minimal={false}
    modifiers={Object {}}
    openOnTargetFocus={true}
    position="bottom"
    targetTagName="span"
    transitionDuration={300}
    usePortal={true}
    wrapperTagName="span"
  >
    <Blueprint3.Button
      icon="applications"
      text="Editors"
    />
  </Blueprint3.Popover>
</Fragment>
`;

exports[`EditorDropdown component renders the extra button if the FIDDLE_DOCS_DEMOS is set 1`] = `
<Fragment>
  <Blueprint3.Popover
    boundary="scrollParent"
    captureDismiss={false}
    content={
      <Blueprint3.Menu>
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="index.html"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="HTML (index.html)"
        />
        <Blueprint3.MenuItem
          disabled={false}
          icon="eye-open"
          id="renderer.js"
          multiline={false}
          onClick={[Function]}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="Renderer Process (renderer.js)"
        />
        <React.Fragment>
          <Blueprint3.MenuDivider />
          <Blueprint3.MenuItem
            disabled={false}
            icon="plus"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Add Custom Editor"
          />
        </React.Fragment>
        <React.Fragment>
          <Blueprint3.MenuDivider />
          <Blueprint3.MenuItem
            disabled={false}
            icon="grid-view"
            multiline={false}
            onClick={[Function]}
            popoverProps={Object {}}
            shouldDismissPopover={true}
            text="Reset Layout"
          />
        </React.Fragment>
      </Blueprint3.Menu>
    }
    defaultIsOpen={false}
    disabled={false}
    fill={false}
    hasBackdrop={false}
    hoverCloseDelay={300}
    hoverOpenDelay={150}
    inheritDarkTheme={true}
    interactionKind="click"
    minimal={false}
    modifiers={Object {}}
    openOnTargetFocus={true}
    position="bottom"
    targetTagName="span"
    transitionDuration={300}
    usePortal={true}
    wrapperTagName="span"
  >
    <Blueprint3.Button
      icon="applications"
      text="Editors"
    />
  </Blueprint3.Popover>
</Fragment>
`;
